rye/src/cli/publish.rs

```diff
@@ -0,0 +1,210 @@
+use std::io::{Read, Write};
+use std::path::PathBuf;
+use std::process::{Command, Stdio};
+
+use age::{
+    secrecy::{ExposeSecret, Secret},
+    Decryptor, Encryptor,
+};
+use anyhow::{bail, Context, Error};
+use clap::Parser;
+use rpassword::read_password;
+use toml_edit::{Item, Table};
+
+use crate::bootstrap::ensure_self_venv;
+use crate::config::{get_credentials, write_credentials};
+use crate::pyproject::PyProject;
+use crate::utils::CommandOutput;
+
+/// Publish packages to a package repository.
+#[derive(Parser, Debug)]
+pub struct Args {
+    /// The distribution files to upload to the repository (defaults to <workspace-root>/dist/*).
+    dist: Option<Vec<PathBuf>>,
+    /// The repository to publish to (defaults to 'pypi').
+    #[arg(short, long, default_value = "pypi")]
+    repository: String,
+    /// The repository url to publish to (defaults to https://upload.pypi.org/legacy/).
+    #[arg(long, default_value = "https://upload.pypi.org/legacy/")]
+    repository_url: String,
+    /// An access token used for the upload.
+    #[arg(long)]
+    token: Option<String>,
+    /// Sign files to upload using GPG.
+    #[arg(long)]
+    sign: bool,
+    /// GPG identity used to sign files.
+    #[arg(short, long)]
+    identity: Option<String>,
+    /// Path to alternate CA bundle.
+    #[arg(long)]
+    cert: Option<PathBuf>,
+    /// Enables verbose diagnostics.
+    #[arg(short, long)]
+    verbose: bool,
+    /// Turns off all output.
+    #[arg(short, long, conflicts_with = "verbose")]
+    quiet: bool,
+}
+
+pub fn execute(cmd: Args) -> Result<(), Error> {
+    let output = CommandOutput::from_quiet_and_verbose(cmd.quiet, cmd.verbose);
+    let venv = ensure_self_venv(output)?;
+    let project = PyProject::discover()?;
+
+    // Get the files to publish.
+    let files = match cmd.dist {
+        Some(paths) => paths,
+        None => vec![project.workspace_path().join("dist").join("*")],
+    };
+
+    // a. Get token from arguments and offer encryption, then store in credentials file.
+    // b. Get token from ~/.rye/credentials keyed by provided repository and provide decryption option.
+    // c. Otherwise prompt for token and provide encryption option, storing the result in credentials.
+    let repository = &cmd.repository;
+
+    // If -r is pypi but the url isn't pypi then bail
+    if repository == "pypi"
+        && !matches!(
+            cmd.repository_url.trim_end_matches('/'),
+            "https://upload.pypi.org/legacy" | "https://upload.pypi.org"
+        )
+    {
+        bail!("invalid pypi url {} (use -h for help)", cmd.repository_url);
+    }
+
+    let mut credentials = get_credentials()?;
+    credentials
+        .entry(repository)
+        .or_insert(Item::Table(Table::new()));
+
+    let token = if let Some(token) = cmd.token {
+        let secret = Secret::new(token);
+        let maybe_encrypted = prompt_maybe_encrypt(&secret)?;
+        let encoded = hex::encode(maybe_encrypted.expose_secret());
```

Comment:
Could just encode the encrypted bytes. If its not encrypted we could just store it as-is.

Reply:
Added in abd82b2e2900253d2887c3ad3395853d7432b0a7